apps/emqx_plugins/src/emqx_plugins_serde.erl

```diff
@@ -0,0 +1,275 @@
+    ok.
+
+%%-------------------------------------------------------------------------------------------------
+%% Internal fns
+%%-------------------------------------------------------------------------------------------------
+
+read_plugin_avsc() ->
+    Pattern = filename:join([emqx_plugins:install_dir(), "*", "config_schema.avsc"]),
+    lists:foldl(
+        fun(AvscPath, AccIn) ->
+            case read_avsc_file(AvscPath) of
+                {ok, Avsc} ->
+                    [_, NameVsn | _] = lists:reverse(filename:split(AvscPath)),
+                    AccIn#{to_bin(NameVsn) => Avsc};
+                {error, Reason} ->
+                    ?SLOG(warning, Reason),
+                    AccIn
+            end
+        end,
+        _Acc0 = #{},
+        filelib:wildcard(Pattern)
+    ).
+
+build_serdes(Schemas) ->
+    maps:foreach(fun do_build_serde/2, Schemas).
+
+do_build_serde(NameVsn, Avsc) ->
+    try
+        Serde = make_serde(NameVsn, Avsc),
+        true = ets:insert(?PLUGIN_SERDE_TAB, Serde),
+        ok
+    catch
+        Kind:Error:Stacktrace ->
+            ?SLOG(
+                error,
+                #{
+                    msg => "error_building_plugin_schema_serde",
+                    name => NameVsn,
+                    kind => Kind,
+                    error => Error,
+                    stacktrace => Stacktrace
+                }
+            ),
+            {error, Error}
+    end.
+
+make_serde(NameVsn, Avsc) ->
+    Store0 = avro_schema_store:new([map]),
+    %% import the schema into the map store with an assigned name
+    %% if it's a named schema (e.g. struct), then Name is added as alias
+    Store = avro_schema_store:import_schema_json(NameVsn, Avsc, Store0),
+    #plugin_schema_serde{
+        name = NameVsn,
+        eval_context = Store
+    }.
+
+ensure_serde_absent(Name) when not is_binary(Name) ->
+    ensure_serde_absent(to_bin(Name));
+ensure_serde_absent(Name) ->
+    case lookup_serde(Name) of
+        {ok, _Serde} ->
+            _ = ets:delete(?PLUGIN_SERDE_TAB, Name),
+            ok;
+        {error, not_found} ->
+            ok
+    end.
+
+async_delete_serdes(Names) ->
+    gen_server:cast(?MODULE, {delete_serdes, Names}).
+
+with_serde(WhichOp, Fun) ->
+    try
+        Fun()
+    catch
+        throw:Reason ->
+            ?SLOG(error, Reason#{
+                which_op => WhichOp,
+                reason => emqx_utils:readable_error_msg(Reason)
+            }),
+            {error, Reason};
+        error:Reason:Stacktrace ->
+            %% unexpected errors, log stacktrace
+            ?SLOG(warning, #{
+                msg => "plugin_schema_op_failed",
+                which_op => WhichOp,
+                exception => Reason,
+                stacktrace => Stacktrace
+            }),
+            {error, #{
+                which_op => WhichOp,
+                reason => Reason
+            }}
+    end.
+
+eval_serde_fun(Op, ErrMsg, SerdeName, Args) ->
+    fun() ->
```

Comment:
don't think it's necessary to return a func here,
there is no laziness inside this fun, just pass all the args to with_serde should be more straightforward.